rust/repositories.bzl

```diff
@@ -42,98 +55,272 @@ filegroup(
+        target_triple = target_triple,
+        tool_name = "rustc",
+        tool_subdirectory = "rustc",
+        version = ctx.attr.version,
+    )
+
+    compiler_BUILD = BUILD_for_compiler(target_triple)
+
+    return compiler_BUILD
+
+def load_rust_stdlib(ctx, target_triple):
+    """Loads a rust standard library and yields corresponding BUILD for it
+
+    Args:
+      ctx: A repository_ctx.
+      target_triple: The rust-style target triple of the tool
+    Returns:
+      The BUILD file contents for this stdlib, and a toolchain decl to match
+    """
+
+    load_arbitrary_tool(
+        ctx,
+        iso_date = ctx.attr.iso_date,
+        param_prefix = "rust-std_",
+        target_triple = target_triple,
+        tool_name = "rust-std",
+        tool_subdirectory = "rust-std-{}".format(target_triple),
+        version = ctx.attr.version,
+    )
+
+    stdlib_BUILD = BUILD_for_stdlib(target_triple)
+    toolchain_BUILD = BUILD_for_toolchain(
+        name = "toolchain_for_{}".format(target_triple),
+        exec_triple = ctx.attr.exec_triple,
+        target_triple = target_triple,
+        workspace_name = ctx.attr.name,
+    )
+
+    return stdlib_BUILD + toolchain_BUILD
+
+def _rust_toolchain_repository_impl(ctx):
+    """The implementation of the rust toolchain repository rule."""
+
+    _check_version_valid(ctx.attr.version, ctx.attr.iso_date)
+
+    BUILD_components = [
+        load_rust_compiler(ctx),
+        load_rust_stdlib(ctx, ctx.attr.exec_triple),
     ]
 
-    all_toolchain_BUILDs = []
-    for toolchain in all_toolchains:
-        all_toolchain_BUILDs.append(BUILD_for_toolchain(toolchain[0], toolchain[1]))
+    for extra_stdlib_triple in ctx.attr.extra_stdlib_triples:
+        BUILD_components.append(load_rust_stdlib(ctx, extra_stdlib_triple))
 
-    return """
-load("@io_bazel_rules_rust//rust:toolchain.bzl", "rust_toolchain")
+    ctx.file("WORKSPACE", "")
+    ctx.file("BUILD", "\n".join(BUILD_components))
+
+rust_toolchain_repositories = repository_rule(
+    attrs = {
+        "version": attr.string(mandatory = True),
+        "iso_date": attr.string(),
+        "exec_triple": attr.string(mandatory = True),
+        "extra_stdlib_triples": attr.string_list(),
+    },
+    implementation = _rust_toolchain_repository_impl,
+)
+
+def rust_repository_set(name, version, exec_triple, extra_stdlib_triples, iso_date = None):
+    """Assembles a remote repository for the given params and yielding the 
+    names of the generated toolchains.
+
+    Args:
+      name: The name of the generated repository
+      version: The version of the tool among "nightly", "beta', or an exact version.
+      iso_date: The date of the tool (or None, if the version is a specific version).
+      exec_triple: The rust-style target that this compiler runs on
+      extra_stdlib_triples: Additional rust-style targets that this set of toolchains
+                            should support.
+    """
+
+    rust_toolchain_repositories(
```

Comment:
can this return the inner labels, or can we make a constant for "toolchain_for_{triple}"?

Reply:
This is a repository_rule, so I don't think it can return a conventional value. We just have to have a contract between the implementation of the repository rule and the macro that wraps it.

How does an (optional) "toolchain_name_prefix" arg sound to make the connection between the distinct usages of "toolchain_for" clearer? I made the change already, LMK what you think.